Utility scale QAOA tutorial

miamico opened the pull request:
Adds a new tutorial for QAOA with code for running experiment at utility scale. All dependencies from Application modules are removed

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/Qiskit/documentation?pullRequest=1238) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/Qiskit/documentation?pullRequest=1238) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/Qiskit/documentation?pullRequest=1238) it.</sub>

abbycross commented:
Happy to do style/language review when this is ready to go. Just lmk!

miamico commented:
posting here some comments received through other channels:
- before digging into the tutorial, I think it would help a ton to motivate the problem a bit more. You mention some applications, which is good, but I think it’s also worth mentioning that it’s an NP hard problem and then comparing the best classical algorithm to the quantum algorithm. At the end of the tutorial you say that current processors are noisy so the quantum algorithm doesn’t offer an advantage. But how much better can they do in principle? If you just leave it at, “quantum computers can do no better than classical,” then the reader will wonder why they’re bothering with learning this.

- I find the starting formulation of the graph problem → QUBO → hamiltonian mapping a little hard to follow and also maybe unnecessarily convoluted? If you formulate the initial graph problem a little differently (as Jochen Rau does in [this video](https://www.youtube.com/watch?v=LMmQ5LD308A)) can’t you just map directly from the graph cost function to a Hamiltonian? If there’s something wrong with this approach and you’d like to stick with what you’re doing, I would suggest explaining the steps a little more and maybe even avoiding ever writing “QUBO.” It sounds intimidating and the term itself is entirely unnecessary for the learner’s understanding of the problem at hand.

- Update to primitives V2

javabster commented on tutorials/qaoa-utility-scale/data/125node_eagle_depth_one.qpy:
Are all these .qpy files necessary? Users should be able to download and run the notebooks, so any additional data and utility functions should be put into the notebook itself. afaik we dont have the ability on the learning platform to handle additional data files like this

javabster commented on tutorials/qaoa-utility-scale/imgs/patterns.png:
is this image still being used or can it be removed from this PR?

miamico replied:
removed

javabster commented on tutorials/qaoa-utility-scale/qopt_example.ipynb:
Can we come up with a more descriptive title for this tutorial? It should be something specific (and ideally enticing) for a reader to want to click on it

javabster replied:
Can you check that the headings for the pattern steps match what is used for other tutorials?

javabster replied:
In general I think this notebook could use a grammar and spelling check (cc @abbycross 😄 )

javabster replied:
I think it would be nice to call out in the intro section of this tutorial that you can run it within the 10 minute limit of the open plan (also make sure to check that is actually true, I haven't personally validated that claim 😅 )

javabster replied:
For the introduction to step 2 when you're giving the TL;DR about transpilation I think it would be good to direct people towards some of the docs.quantum.ibm.com/tranpile resources for more detail

javabster replied:
Also in the step 2 intro - why is a simulator used? we should always show a real device unless absolutely necessary to do otherwise. Also I'm generally not really sure what the purpose of that code snippet is for, if it is just to show what transpilation is I think we can cut it and instead add some text encouraging people to go to the docs if they dont know what transpilation is

javabster replied:
step 3 also uses a simulator - it should use real hardware pls

javabster replied:
part 2 has a bunch of "call to action" text that i'm assuming was used as part of the in-person workshop and should probably be removed for the tutorial format

javabster replied:
first code cell in the part 2 has a bunch of code comments for eagle and heron - we should probably just pick one and stick with it. Or explain that we're going to compare the results later

javabster replied:
in the scale up section the first diagram of the nodes looks janky 😅

javabster replied:
in the scale up step 2 section we mention different optimization strategies like the AI transpiler passes. If we're going to mention that stuff but not show a code example in the notebook then we should add a link to the docs where they can go find more info

javabster replied:
in scale up step 3 I think we should remove the explanation about loading saved data and only show the code cells that show users how to do it themselves. Mostly because we dont have a way in LP to load saved data like this

javabster replied:
would be nice if you can explain the graphs in the final step a bit more, rather than just saying "heres a cumulative distribution"

miamico replied:
> Can we come up with a more descriptive title for this tutorial? It should be something specific (and ideally enticing) for a reader to want to click on it

I'm giving another stab but maybe someone with more imagination can help me?

miamico replied:
> Also in the step 2 intro - why is a simulator used? we should always show a real device unless absolutely necessary to do otherwise. Also I'm generally not really sure what the purpose of that code snippet is for, if it is just to show what transpilation is I think we can cut it and instead add some text encouraging people to go to the docs if they dont know what transpilation is

- Simulator was use for ease of usage, we can definitely switch to hardware.
- We need to transpile the circuit to the hardware before execution (the sampler/estimator don't do that anymore). Step 2 seems the most natural place to put the transpilation step

miamico replied:
> in scale up step 3 I think we should remove the explanation about loading saved data and only show the code cells that show users how to do it themselves. Mostly because we dont have a way in LP to load saved data like this

Ok, let's do this as a last step to avoid having to keep sending jobs to the device to execute the notebook

miamico replied:
> in the scale up section the first diagram of the nodes looks janky 😅

I need some help with this, have very little idea of plotting with rustworkx

frankharkins replied:
> > in the scale up section the first diagram of the nodes looks janky 😅
>
> I need some help with this, have very little idea of plotting with rustworkx

It doesn't seem like there's much control over it. If there's some way to get the coordinates they use for the graphs on quantum.ibm.com then we can input them into the drawer, but I imagine that'll convolute the tutorial.

I don't think it's too bad at the moment so might be best leaving it.

javabster reviewed (commented):
Left a bunch of comments but it's getting there! thanks @miamico and @kaelynj 😄

abbycross commented:
Github won't let me suggest edits because of the size of the notebook file, so I am making some IBM style updates (such as rewording not to use "we") locally, and will merge to this branch as soon as I've gotten through it!

abbycross commented:
Looking at Pattern step 2 in the text: we should add a link to the docs for transpiler passes. Is this the right link to add? https://docs.quantum.ibm.com/api/qiskit/transpiler_passes#transpiler-passes -- or instead, something in the platform docs (https://docs.quantum.ibm.com/transpile)? cc: @kaelynj 
<img width="1317" alt="Screenshot 2024-05-20 at 11 42 27 AM" src="https://github.com/Qiskit/documentation/assets/28769477/dd2610da-31ef-42e0-8fc4-c8d3ab2f8f6a">

abbycross commented:
I am hesitant to "IBM-ify" the style of the Appendices, but also wonder if they belong at the end of the notebook since we don't use appendices anywhere else. Is there a better place to store and display that content? Thoughts @javabster @kaelynj @miamico ?

miamico commented:
I would vote for having appendices as this seems to be a common occurence in the utility notebook I've been working on. Maybe a link to external reference would be enough though

kcmccormibm commented:
Is it possible to have an alternate cell to the COBYLA optimization that open users can use to just load the output of a previous job? As an open user, sessions doesn't work, so I have to wait ~3 hrs for each iteration of this to run. This makes this notebook practically unrunnable for an open user.

javabster commented:
Thanks folks, this is very nearly there! Some last comments from me:

- In the intro can you explain that you'll first run a small scale version of this problem and then scale it up to utility scale? Also in general I think we should follow the same heading structure as other similar tutorials [like this one](https://learning.quantum.ibm.com/tutorial/combine-error-mitigation-options-with-the-estimator-primitive)
- notebook is missing a "requirements" section at the start
- The first section under the "Part 1 (small scale) heading feels like it should be part of the "map" step, no?
- in general can you double check that the pattern headings are the same as the other recent tutorials we've shipped
- small jargon thing - I've heard from leadership/marketing that we're supposed to say "AI enhanced transpiler passes" not "AI transpiler passes"
- As the appendices are so small I think it would be better to just include them directly in the tutorial where it's most relevant.

Lastly, @kcmccormibm has a fair point, this tutorial in particular is supposed to be accessible for open users. The original notebook was runnable within the 10min QPU time available to open users, is that still the case after all the edits that have been made? I'm not sure there's much we can do about the classical optimization part taking a long time though 🤔 does anyone have any suggestions for how to improve that?

kcmccormibm commented:
Hi! I just wanted to clarify that I don't *think* the QPU time is a problem (although I haven't been able to fully run it to know for sure) -- for open users I think the main pain point will be the time in the queue. Waiting several hours in the queue between each iteration of the COBYLA optimization does not seem great, but then again that's probably unavoidable...

kaelynj commented:
> Thanks folks, this is very nearly there! Some last comments from me:
> *  In the intro can you explain that you'll first run a small scale version of this problem and then scale it up to utility scale? Also in general I think we should follow the same heading structure as other similar tutorials [like this one](https://learning.quantum.ibm.com/tutorial/combine-error-mitigation-options-with-the-estimator-primitive)
> 
> * notebook is missing a "requirements" section at the start
> 
> * The first section under the "Part 1 (small scale) heading feels like it should be part of the "map" step, no?
> 
> * in general can you double check that the pattern headings are the same as the other recent tutorials we've shipped
> 
> * also can we add an estimated QPU time at the top like we do for the other notebooks
> 
> * small jargon thing - I've heard from leadership/marketing that we're supposed to say "AI enhanced transpiler passes" not "AI transpiler passes"
> 
> * As the appendices are so small I think it would be better to just include them directly in the tutorial where it's most relevant.

Most of this feedback has now been incorporated. I've moved the content from the appendices into an expanding subsection and double checked the headers. The section header for part 1 and 2 I think should stay separated from the patterns steps headers since we're essentially creating two different patterns workflows (one small scale and another at utility scale) and the reader should be able to follow that.

> Lastly, @kcmccormibm has a fair point, this tutorial in particular is supposed to be accessible for open users. The original notebook was runnable within the 10min QPU time available to open users, is that still the case after all the edits that have been made? I'm not sure there's much we can do about the classical optimization part taking a long time though 🤔 does anyone have any suggestions for how to improve that?

As for this unfortunately I don't think there's much we can do about queue times/executing within a session and as far s I know we don't have the infrastructure set up to keep a pre-baked set of results ready to go for the reader (right @Eric-Arellano?). Also @miamico do you have a rough idea of the QPU time that was needed to run this notebook?

Eric-Arellano commented:
> and as far s I know we don't have the infrastructure set up to keep a pre-baked set of results ready to go for the reader (right @Eric-Arellano?).

The only way I can think toto do that would be hardcoding the results in the Jupyter notebook and doing something like this:

```python
# Reader: change this value to `False` if you want to run on hardware to get
# your own results
USE_CACHED_RESULTS = True


...

if USE_CACHED_RESULTS:
  my_result = ...
else:
   my_result = sampler.run()
```

christopherporter1 commented:
Hi All, this was just shared with me. Fantastic work! A few suggestions:

- I think the current version is missing a "from qiskit import transpile" line right above "candidate_circuit = transpile(circuit, backend, optimization_level=3)"
- The first 5-qubit run often returns non-optimal cuts, even on a noise-free simulator. When I increased reps to 2, and added parameters accordingly, I never again got a non-optimal cut. So maybe that's worth implementing to have a solid small scale example.
- I propose specifying 'tol' and 'maxiter' for COBYLA, although now that I changed the reps, it doesn't seem to matter.
- Is there a reason for sticking with Sherbrooke as opposed to least_busy? No worries if there is something about benchmarking that required a consistent backend. But switching to least_busy saved me hours and hours when checking this.
- There should probably be a deeper QAOA ansatz (more than just reps = 1) for the utility scale job. I'm not proposing very deep, since the goal is to make it runnable in open-plan time. But if the 5 qubit job needs reps = 2, then the large one should at least match that and address in text the need to tailor settings accordingly for utility scale jobs.

Thanks all! Beautiful tutorial!

javabster commented on tutorials/qaoa-utility-scale/data/sampler/qaoa_eagle_97e3c32c.json:
I think this file needs to be removed? we can't load it in the platform as far as I know, all the data should be generated in the notebook itself

javabster commented on tutorials/qaoa-utility-scale/qopt_example.ipynb:
for the requirements section I think we need to mention they should install the qiskit sdk with the visualisation extension. I think that also means you can remove the mpl requirement bc that gets installed as part of the qiskit visualisation install (I _think_)

javabster commented:
also, IMPORTANT: we should not be using `transpile()` anymore, we should be using preset passmangers instead

miamico commented:
I've added the estimate for the QPU time. In my experiments it took 6s per function call of COBYLA (and about 40 calls). I've also switched from `transpile` to the passmanager

kaelynj commented:
Just noticed/fixed a bug in the small scale example where the call to instantiate the sampler wasn't using the session being generated.

I'm also wondering about editing the content from @christopherporter1's feedback:
> * There should probably be a deeper QAOA ansatz (more than just reps = 1) for the utility scale job. I'm not proposing very deep, since the goal is to make it runnable in open-plan time. But if the 5 qubit job needs reps = 2, then the large one should at least match that and address in text the need to tailor settings accordingly for utility scale jobs.

I switched reps=2 for the small scale experiment and introduced a `tol=1e-3` to the minimization loop. Is that a reasonable enough tolerance to set @miamico (especially for the larger scale example)? Or could we get away with a lower tolerance?

kaelynj reviewed (commented):
Added just a few tweaks to the wording of the sections, but otherwise this looks good to me! :rocket: @javabster ready to take one last look?

abbycross commented:
I see some more wording tweaks to make but GitHub won't let me do it here since the file is so big - so I will make the changes locally and put them into one commit.

abbycross commented:
Had some merge conflict problems, so another commit is coming from me - stand by before merging, please

nathanearnestnoble commented:
Just running through the notebook and noticed deprecation warning -- perhaps we should update the session formatting? 

`DeprecationWarning: The 'session' keyword arguments are deprecated as of qiskit-ibm-runtime 0.23.0 and will be removed no sooner than 3 months after the release date. Please use the 'mode' parameter instead.
  estimator = Estimator(session=session)`

javabster commented:
this is looking good to me, I just have one question - I see a lot of references to quantum annealing and im just wondering how relevant it is to mention? I personally don't know much about quantum annealing so it doesn't really mean much to me, and it is somewhat confusing bc we are a superconducting qubit company not an annealing one so the reference feels odd. Unless it is super important for the reader's understanding then I would just remove those references. Also there is a typo in the very last sentence in the notebook (should be "around" not aroud")

kaelynj commented:
> this is looking good to me, I just have one question - I see a lot of references to quantum annealing and im just wondering how relevant it is to mention? I personally don't know much about quantum annealing so it doesn't really mean much to me, and it is somewhat confusing bc we are a superconducting qubit company not an annealing one so the reference feels odd. Unless it is super important for the reader's understanding then I would just remove those references. Also there is a typo in the very last sentence in the notebook (should be "around" not aroud")

I think it's helpful context to mention annealing here. QAOA is inspired/based on quantum annealing, but is adapted for a gate-based quantum computer and the only references I see to annealing provide that context to the reader.

kaelynj commented:
> Just running through the notebook and noticed deprecation warning -- perhaps we should update the session formatting?
> 
> `DeprecationWarning: The 'session' keyword arguments are deprecated as of qiskit-ibm-runtime 0.23.0 and will be removed no sooner than 3 months after the release date. Please use the 'mode' parameter instead. estimator = Estimator(session=session)`

With regard to this, I think the `mode` argument was only *just* released this month and so users may not be running on the very latest version. We could simply add in a comment within the relevant code cells to mention that if they using `qiskit-ibm-runtime>0.23.0` they should use the `mode` argument.

kaelynj reviewed (approved):
Switched up the primitive calls to use `mode` and left a note to users to switch to either `backend=` or `session=` if they are using `qiskit-ibm-runtime<0.24.0`. Also added the notebook to the CI.

Everything looks good to me now!

Eric-Arellano commented:
Force merging because the CLA check is broken and this is failing due to being on fork, even though it doesn't need to fail (https://github.com/Qiskit/documentation/issues/1649).
